File: java/codegen/src/main/java/io/yosina/codegen/generators/CombinedTransliteratorGenerator.java
package io.yosina.codegen.generators;

import io.yosina.codegen.Artifact;
import io.yosina.codegen.UnicodeUtils;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/** Generates the combined transliterator that maps single characters to arrays of characters. */
public class CombinedTransliteratorGenerator implements TransliteratorGenerator {
    private static final String COMBINED_TRANSLITERATOR_TEMPLATE =
            """
package io.yosina.transliterators;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.io.IOException;
import java.io.InputStream;
import java.io.UncheckedIOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

import io.yosina.Char;
import io.yosina.CharIterator;
import io.yosina.CodePointTuple;
import io.yosina.Transliterator;
import io.yosina.annotations.RegisteredTransliterator;

/**
 * Auto-generated transliterator for %2$s.
 * Replace single characters with arrays of characters.
 */
@RegisteredTransliterator(name = "%2$s")
public class %1$s implements Transliterator {
    private static final Map<CodePointTuple, int[]> mappings;

    static {
        final Map<CodePointTuple, int[]> mappings_ = new TreeMap<>();
        final ByteBuffer b;
        try {
            try (final InputStream s = %1$s.class.getResourceAsStream("%2$s.data")) {
                b = ByteBuffer.wrap(s.readAllBytes()).order(ByteOrder.BIG_ENDIAN);
            }
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
        while (b.hasRemaining()) {
            final int key1 = b.getInt(), key2 = b.getInt();
            final int length = b.getInt();
            final int[] value = new int[length];
            for (int i = 0; i < length; i++) {
                value[i] = b.getInt();
            }
            mappings_.put(CodePointTuple.of(key1, key2), value);
        }
        mappings = Collections.unmodifiableMap(mappings_);
    }

    private static class %1$sCharIterator implements CharIterator {
        private final CharIterator input;
        private final Map<CodePointTuple, int[]> mappings;
        private final List<Char> queue = new ArrayList<>();
        private int queueIndex = 0;

        public %1$sCharIterator(CharIterator input, Map<CodePointTuple, int[]> mappings) {
            this.input = input;
            this.mappings = mappings;
        }

        @Override
        public Char next() {
            // Return queued characters first
            if (queueIndex < queue.size()) {
                return queue.get(queueIndex++);
            }

            // Reset queue for new character
            queue.clear();
            queueIndex = 0;

            if (!input.hasNext()) {
                return null;
            }

            Char ch = input.next();
            if (ch.isSentinel()) {
                return ch;
            }

            int[] replacement = mappings.get(ch.get());
            if (replacement != null) {
                // Create new characters for each replacement
                for (int i = 0; i < replacement.length; i++) {
                    queue.add(new Char(
                        CodePointTuple.of(replacement[i], -1),
                        ch.getOffset() + i,
                        ch
                    ));
                }
                return queue.get(queueIndex++);
            } else {
                return ch;
            }
        }

        @Override
        public boolean hasNext() {
            return queueIndex < queue.size() || input.hasNext();
        }
    }

    @Override
    public CharIterator transliterate(CharIterator input) {
        return new %1$sCharIterator(input, mappings);
    }

    /** Creates a new Combined transliterator. */
    public %1$s() {
    }
}
""";

    private final Map<String, String> mappings;
    private final String className;
    private final String name;

    public CombinedTransliteratorGenerator(
            Map<String, String> mappings, String className, String name) {
        this.mappings = mappings;
        this.className = className;
        this.name = name;
    }

    @Override
    public List<Artifact> generate() {
        List<Artifact> artifacts = new ArrayList<>();

        // Generate source file
        artifacts.add(
                new Artifact(
                        Artifact.Type.SOURCE,
                        Path.of(String.format("%s.java", className)),
                        ByteBuffer.wrap(
                                String.format(COMBINED_TRANSLITERATOR_TEMPLATE, className, name)
                                        .getBytes(StandardCharsets.UTF_8))));

        // Generate binary data file
        ByteBuffer dataBuffer = generateBinaryData();
        artifacts.add(new Artifact(Artifact.Type.RESOURCE, Path.of(name + ".data"), dataBuffer));

        return artifacts;
    }

    private ByteBuffer generateBinaryData() {
        // Calculate total size needed
        int totalSize = 0;
        for (Map.Entry<String, String> entry : mappings.entrySet()) {
            // 2 ints for key + 1 int for length + n ints for values
            totalSize += 3 * 4 + entry.getValue().codePointCount(0, entry.getValue().length()) * 4;
        }

        ByteBuffer buffer = ByteBuffer.allocate(totalSize).order(ByteOrder.BIG_ENDIAN);

        for (Map.Entry<String, String> entry : mappings.entrySet()) {
            int fromCodePoint = UnicodeUtils.parseUnicodeCodepoint(entry.getKey());
            String toStr = entry.getValue();

            // Write key (code point tuple)
            buffer.putInt(fromCodePoint);
            buffer.putInt(-1); // Second code point is always -1 for single characters

            // Write value array
            int[] toCodePoints = toStr.codePoints().toArray();
            buffer.putInt(toCodePoints.length);
            for (int codePoint : toCodePoints) {
                buffer.putInt(codePoint);
            }
        }

        return buffer.rewind();
    }
}

File: csharp/src/Yosina.Codegen/Generators/CombinedTransliteratorGenerator.cs
// Copyright (c) Yosina. All rights reserved.

namespace Yosina.Codegen.Generators;

/// <summary>
/// Generates source code for the CombinedTransliterator that replaces single characters with arrays of characters.
/// </summary>
public static class CombinedTransliteratorGenerator
{
    /// <summary>
    /// Generates the source code for the CombinedTransliterator class.
    /// </summary>
    /// <param name="mappings">Dictionary mapping source code points to arrays of replacement characters.</param>
    /// <returns>The generated C# source code as a string.</returns>
    public static string Generate(IDictionary<int, string[]> mappings)
    {
        var className = "CombinedTransliterator";
        var registrationName = "combined";
        var description = "Replace single characters with arrays of characters.";

        var mappingEntries = mappings
            .OrderBy(m => m.Key)
            .Select(mapping => FormatMappingEntry(mapping.Key, mapping.Value));

        var mappingsDict = string.Join("\n", mappingEntries);

        return GenerateTemplate(className, registrationName, description, mappingsDict);
    }

    private static string FormatMappingEntry(int fromCodePoint, string[] toChars)
    {
        var charArray = string.Join(", ", toChars.Select(c => $"0x{char.ConvertToUtf32(c, 0):X}"));
        return $"        [0x{fromCodePoint:X}] = new int[] {{ {charArray} }},";
    }

    private static string GenerateTemplate(string className, string registrationName, string description, string mappingsDict)
    {
        return $@"namespace Yosina.Transliterators;

/// <summary>
/// {description}
/// This file is auto-generated. Do not edit manually.
/// </summary>
[RegisteredTransliterator(""{registrationName}"")]
public class {className} : ITransliterator
{{
    private static readonly Dictionary<int, int[]> Mappings = new()
    {{
{mappingsDict}
    }};

    public IEnumerable<Character> Transliterate(IEnumerable<Character> input)
    {{
        return new CombinedCharEnumerator(input.GetEnumerator(), Mappings);
    }}

    private class CombinedCharEnumerator : IEnumerable<Character>, IEnumerator<Character>
    {{
        private readonly IEnumerator<Character> input;
        private readonly Dictionary<int, int[]> mappings;
        private int offset;
        private bool disposed;
        private readonly Queue<Character> queue = new();

        public CombinedCharEnumerator(IEnumerator<Character> input, Dictionary<int, int[]> mappings)
        {{
            this.input = input;
            this.mappings = mappings;
        }}

        public Character Current {{ get; private set; }} = null!;

        object System.Collections.IEnumerator.Current => this.Current;

        public bool MoveNext()
        {{
            if (this.disposed)
            {{
                return false;
            }}

            // If we have queued characters from a previous expansion, return them first
            if (this.queue.Count > 0)
            {{
                this.Current = this.queue.Dequeue();
                return true;
            }}

            if (!this.input.MoveNext())
            {{
                return false;
            }}

            var c = this.input.Current;
            if (c.IsSentinel)
            {{
                this.Current = c.WithOffset(this.offset);
                return true;
            }}

            if (c.CodePoint.Size() == 1 && this.mappings.TryGetValue(c.CodePoint.First, out var mapped))
            {{
                // Create characters for each mapped code point
                var chars = mapped.Select((codePoint, index) =>
                    new Character((codePoint, -1), this.offset + index, c)).ToArray();

                if (chars.Length > 0)
                {{
                    this.Current = chars[0];
                    this.offset += chars[0].CharCount;

                    // Queue the remaining characters
                    for (int i = 1; i < chars.Length; i++)
                    {{
                        this.queue.Enqueue(chars[i].WithOffset(this.offset));
                        this.offset += chars[i].CharCount;
                    }}
                }}
                else
                {{
                    this.Current = c.WithOffset(this.offset);
                    this.offset += this.Current.CharCount;
                }}
            }}
            else
            {{
                this.Current = c.WithOffset(this.offset);
                this.offset += this.Current.CharCount;
            }}

            return true;
        }}

        public void Reset()
        {{
            this.input.Reset();
            this.offset = 0;
            this.queue.Clear();
        }}

        public void Dispose()
        {{
            if (!this.disposed)
            {{
                this.input.Dispose();
                this.disposed = true;
            }}
        }}

        public System.Collections.Generic.IEnumerator<Character> GetEnumerator() => this;

        System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() => this;
    }}
}}";
    }
}
